src/ndarray/ndarray_function.cu

```diff
@@ -88,5 +94,110 @@ void Copy<gpu, gpu>(const TBlob &from, TBlob *to,
                         s->stream_);
   }
 }
+
+/*!
+ * \brief GPU impl of elemwise sum for rowsparse tensors.
+ */
+void ElementwiseSumRspImpl(mshadow::Stream<gpu>* s,
+                       const std::vector<NDArray>& nds,
+                       NDArray* out) {
+  using namespace mxnet::op;
+  using namespace rowsparse;
+  using nnvm::dim_t;
+  CHECK_EQ(out->storage_type(), kRowSparseStorage)
+    << "Expected rowsparse storage_type (" << out->storage_type() << " given)";
+  int init = 0;
+  for (const auto& nd : nds) {
+    if (nd.storage_initialized()) {
+      init++;
+    }
+  }
+  if (init == 0) {
+    FillZerosRspImpl<gpu>(s, out);
+    return;
+  }
+  const dim_t num_rows = out->shape()[0];
+  const dim_t row_length = out->shape().ProdShape(1, out->shape().ndim());
+  // TODO(stefan): use temporary workspace from OpContext instead of cudaMalloc
+  MSHADOW_TYPE_SWITCH(out->dtype(), DType, {  // data type
+    MSHADOW_IDX_TYPE_SWITCH(out->aux_type(kIdx), IType, {  // row_idx type
+      // Allocate temporary array row_flg
+      IType* row_flg = NULL;
+      CUDA_CALL(cudaMalloc(&row_flg, num_rows*sizeof(IType)));
+      dim_t num_threads = num_rows;
+      mxnet_op::Kernel<mxnet_op::set_zero, gpu>::Launch(s, num_threads, row_flg);
+      // Mark row_flg array with 1s for non-zero rows
+      for (const auto& nd : nds) {
+        if (nd.storage_initialized()) {
+          const IType* nd_row_idx = nd.aux_data(kIdx).dptr<IType>();
+          const dim_t nd_nnr = nd.storage_shape()[0];
+          num_threads = nd_nnr;
+          mxnet_op::Kernel<MarkRspRowFlgKernel, gpu>::Launch(s, num_threads,
+              row_flg, nd_row_idx, nd_nnr);
+        }
+      }
+      // Compute inclusive prefix sum over row_flg
+      void* d_temp_storage = NULL;
+      size_t temp_storage_bytes = 0;
+      cub::DeviceScan::InclusiveSum(d_temp_storage,
+                                    temp_storage_bytes,
+                                    row_flg,
+                                    row_flg,
+                                    num_rows,
+                                    mshadow::Stream<gpu>::GetStream(s));
+      CUDA_CALL(cudaMalloc(&d_temp_storage, temp_storage_bytes));
```

Comment:
If an exception is thrown in a subsequent CUDA_CALL(), is this memory cleaned up?